hbase-server/src/main/java/org/apache/hadoop/hbase/mob/RSMobFileCleanerChore.java

```diff
@@ -0,0 +1,262 @@
+
+    TableDescriptors htds = rs.getTableDescriptors();
+    try {
+      FileSystem fs = FileSystem.get(rs.getConfiguration());
+
+      Map<String, TableDescriptor> map = null;
+      try {
+        map = htds.getAll();
+      } catch (IOException e) {
+        LOG.error("MobFileCleanerChore failed", e);
+        return;
+      }
+      Map<String, Map<String, List<String>>> referencedMOBs = new HashMap<>();
+      for (TableDescriptor htd : map.values()) {
+        // Now clean obsolete files for a table
+        LOG.info("Cleaning obsolete MOB files from table={}", htd.getTableName());
+        List<ColumnFamilyDescriptor> list = MobUtils.getMobColumnFamilies(htd);
+        List<HRegion> regions = rs.getRegions(htd.getTableName());
+        for (HRegion region : regions) {
+          for (ColumnFamilyDescriptor hcd : list) {
+            HStore store = region.getStore(hcd.getName());
+            Collection<HStoreFile> sfs = store.getStorefiles();
+            Set<String> regionMobs = new HashSet<String>();
+            Path currentPath = null;
+            try {
+              //collectinng referenced MOBs
+              for (HStoreFile sf : sfs) {
+                currentPath = sf.getPath();
+                sf.initReader();
+                byte[] mobRefData = sf.getMetadataValue(HStoreFile.MOB_FILE_REFS);
+                byte[] bulkloadMarkerData = sf.getMetadataValue(HStoreFile.BULKLOAD_TASK_KEY);
+                // close store file to avoid memory leaks
+                sf.closeStoreFile(true);
+                if (mobRefData == null) {
+                  if (bulkloadMarkerData == null) {
+                    LOG.warn("Found old store file with no MOB_FILE_REFS: {} - "
+                      + "can not proceed until all old files will be MOB-compacted.", currentPath);
+                    return;
+                  } else {
+                    LOG.debug("Skipping file without MOB references (bulkloaded file):{}",
+                      currentPath);
+                    continue;
+                  }
+                }
+                // file may or may not have MOB references, but was created by the distributed
+                // mob compaction code.
+                try {
+                  SetMultimap<TableName, String> mobs =
+                    MobUtils.deserializeMobFileRefs(mobRefData).build();
+                  LOG.debug("Found {} mob references for store={}", mobs.size(), sf);
+                  LOG.trace("Specific mob references found for store={} : {}", sf, mobs);
+                  regionMobs.addAll(mobs.values());
+                } catch (RuntimeException exception) {
+                  throw new IOException("failure getting mob references for hfile " + sf,
+                    exception);
+                }
+              }
+              //collecting files, MOB included currently being written
+              regionMobs.addAll(
+                store.getStoreFilesBeingWritten().stream().map(path -> path.getName())
```

Comment:
I guess this here is because we need to add all currently being written files because we have no means to know if those have mob refs or not.

Reply:
It is here because mob compaction/flush runs as part of non-mob store compaction/flush that created it. And because of how these work the most convenient way to keep track of compacting/flushing mob files was to add them to the list of compacting/flushing non-mob files. So this list has everything. But only part of the list is actually interesting for us at any given time.